In [1]:
import numpy as np
import pandas as pd

In [20]:
__author__ = 'Chris Potts'


class RSA:
    """Implementation of the core Rational Speech Acts model.

    Parameters
    ----------
    lexicon : `np.array` or `pd.DataFrame`
        Messages along the rows, states along the columns.
    prior : array-like
        Same length as the number of colums in `lexicon`.
    costs : array-like
        Same length as the number of rows in `lexicon`.
    alpha : float
        The temperature parameter. Default: 1.0
    """
    def __init__(self, lexicon, prior, costs, alpha=1.0):
        self.lexicon = lexicon
        self.prior = np.array(prior)
        self.costs = np.array(costs)
        self.alpha = alpha

    def literal_listener(self):
        """Literal listener predictions, which corresponds intuitively
        to truth conditions with priors.

        Returns
        -------
        np.array or pd.DataFrame, depending on `self.lexicon`.
        The rows correspond to messages, the columns to states.

        """
        return rownorm(self.lexicon * self.prior)

    def speaker(self):
        """Returns a matrix of pragmatic speaker predictions.

        Returns
        -------
        np.array or pd.DataFrame, depending on `self.lexicon`.
        The rows correspond to states, the columns to states.
        """
        lit = self.literal_listener().T
        utilities = self.alpha * (safelog(lit) + self.costs)
        return rownorm(np.exp(utilities))

    def listener(self):
        """Returns a matrix of pragmatic listener predictions.

        Returns
        -------
        np.array or pd.DataFrame, depending on `self.lexicon`.
        The rows correspond to messages, the columns to states.
        """
        spk = self.speaker().T
        return rownorm(spk * self.prior)


def rownorm(mat):
    """Row normalization of np.array or pd.DataFrame"""
    return (mat.T / mat.sum(axis=1)).T


def safelog(vals):
    """Silence distracting warnings about log(0)."""
    with np.errstate(divide='ignore'):
        return np.log(vals)


if __name__ == '__main__':
    """Examples"""

    from IPython.display import display


    def display_reference_game(mod):
        d = mod.lexicon.copy()
        d['costs'] = mod.costs
        d.loc['prior'] = list(mod.prior) + [""]
        d.loc['alpha'] = [mod.alpha] + [" "] * mod.lexicon.shape[1]
        display(d)


    # Core lexicon:
    msgs = ['шляпа', 'очки', 'усы']
    states = ['r1', 'r2','r3']
    lex = pd.DataFrame([
        [1.0, 1.0, 0.0],
        [0.0, 0.0, 1.0],
        [0.0, 1.0, 0.0]], index=msgs, columns=states)

In [21]:
print("="*70 + "\nEven priors and all-0 message costs\n")
basic_mod = RSA(lexicon=lex, prior=[1/3, 1/3, 1/3], costs=[0.0, 0.0, 0.0])

display_reference_game(basic_mod)

print("\nLiteral listener")
display(basic_mod.literal_listener())

print("\nPragmatic speaker")
display(basic_mod.speaker())

print("\nPragmatic listener")
display(basic_mod.listener())

Even priors and all-0 message costs



r1        r2        r3 costs
шляпа  1.000000       1.0       0.0   0.0
очки   0.000000       0.0       1.0   0.0
усы    0.000000       1.0       0.0   0.0
prior  0.333333  0.333333  0.333333      
alpha  1.000000


Literal listener


r1   r2   r3
шляпа  0.5  0.5  0.0
очки   0.0  0.0  1.0
усы    0.0  1.0  0.0


Pragmatic speaker


шляпа  очки       усы
r1  1.000000   0.0  0.000000
r2  0.333333   0.0  0.666667
r3  0.000000   1.0  0.000000


Pragmatic listener


r1    r2   r3
шляпа  0.75  0.25  0.0
очки   0.00  0.00  1.0
усы    0.00  1.00  0.0

In [17]:
print("="*70 + "\nEven priors, imbalanced message costs\n")
cost_most = RSA(lexicon=lex,  prior=[0.1, 0.45, 0.45], costs=[0.0, 0.0, 0.0])

display_reference_game(cost_most)

print("\nLiteral listener")
display(cost_most.literal_listener())

print("\nPragmatic speaker")
display(cost_most.speaker())

print("\nPragmatic listener")
display(cost_most.listener())

Even priors, imbalanced message costs



r1    r2    r3 costs
шляпа  1.0   1.0   0.0   0.0
очки   0.0   0.0   1.0   0.0
усы    0.0   1.0   0.0   0.0
prior  0.1  0.45  0.45      
alpha  1.0


Literal listener


r1        r2   r3
шляпа  0.181818  0.818182  0.0
очки   0.000000  0.000000  1.0
усы    0.000000  1.000000  0.0


Pragmatic speaker


шляпа  очки   усы
r1   1.00   0.0  0.00
r2   0.45   0.0  0.55
r3   0.00   1.0  0.00


Pragmatic listener


r1        r2   r3
шляпа  0.330579  0.669421  0.0
очки   0.000000  0.000000  1.0
усы    0.000000  1.000000  0.0

# Задание 4

Приведите свой пример референциальной игры с четырьмя объектами (состояниями, r1, r2, r3, r4) и четырьмя высказываниями (m1, m2, m3, m4), в которой при равных исходных вероятностях объектов, alpha = 1, цене всех высказываний равной 0 вычисляется одна или две скалярные импликатуры. Покажите, что импликатуры вычисляются (таблица для прагматического слушающего + комментарий, описывающий импликатуру).

In [ ]:
Референциальная игра: 

In [23]:
msgs = ['география', 'химия', 'алгебра', 'физика']
states = ['1ученик', '2ученик','3ученик','4ученик']
lex = pd.DataFrame([
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 1.0, 1.0, 0.0],
        [0.0, 0.0, 1.0, 0.0],
        [0.0, 0.0, 0.0, 1.0]], index=msgs, columns=states)
print("="*70 + "\nEven priors and all-0 message costs\n")
basic_mod = RSA(lexicon=lex, prior=[0.25, 0.25, 0.25, 0.25], costs=[0.0, 0.0, 0.0, 0.0])

display_reference_game(basic_mod)

print("\nLiteral listener")
display(basic_mod.literal_listener())
print("\nPragmatic speaker")
display(basic_mod.speaker())

print("\nPragmatic listener")
display(basic_mod.listener())

Even priors and all-0 message costs



1ученик 2ученик 3ученик 4ученик costs
география     1.00     0.0     0.0     0.0   0.0
химия         0.00     1.0     1.0     0.0   0.0
алгебра       0.00     0.0     1.0     0.0   0.0
физика        0.00     0.0     0.0     1.0   0.0
prior         0.25    0.25    0.25    0.25      
alpha         1.00


Literal listener


1ученик  2ученик  3ученик  4ученик
география      1.0      0.0      0.0      0.0
химия          0.0      0.5      0.5      0.0
алгебра        0.0      0.0      1.0      0.0
физика         0.0      0.0      0.0      1.0


Pragmatic speaker


география     химия   алгебра  физика
1ученик        1.0  0.000000  0.000000     0.0
2ученик        0.0  1.000000  0.000000     0.0
3ученик        0.0  0.333333  0.666667     0.0
4ученик        0.0  0.000000  0.000000     1.0


Pragmatic listener


1ученик  2ученик  3ученик  4ученик
география      1.0     0.00     0.00      0.0
химия          0.0     0.75     0.25      0.0
алгебра        0.0     0.00     1.00      0.0
физика         0.0     0.00     0.00      1.0

Обратимся к pragmatic speaker: шкала информативности предмета "химия" дробится на 1/2 и 1/2.Химия нравится ученику 2 и ученику 3, однако третьему ученику нравится еще и алгебра. Можно сделать вывод, что "алгебра" более информативна для третьего ученика. 

# Как меняются вероятности для прагматического слушающего с изменением параметра alpha (оптимальность, прагматичность, рациональность говорящего), параметра цены высказывания, с изменением исходных (априорных) вероятностей объектов? При каких значениях параметров импликатура (или импликатуры, если в Вашей игре их несколько) исчезает?

Let's try to change alpha! Cначала изменим альфу не на очень много, а потом попробуем изменить побольше:

In [29]:
alpha_mod = RSA(lexicon=lex, prior=[0.25, 0.25, 0.25, 0.25], costs=[0.0, 0.0, 0.0, 0.0], alpha=1.5)
print("="*70 + "\nEven priors and all-0 message costs; alpha = 2\n")

display_reference_game(alpha_mod)

print("\nLiteral listener")
display(alpha_mod.literal_listener())

print("\nPragmatic speaker")
display(alpha_mod.speaker())

print("\nPragmatic listener")
display(alpha_mod.listener())

Even priors and all-0 message costs; alpha = 2



1ученик 2ученик 3ученик 4ученик costs
география     1.00     0.0     0.0     0.0   0.0
химия         0.00     1.0     1.0     0.0   0.0
алгебра       0.00     0.0     1.0     0.0   0.0
физика        0.00     0.0     0.0     1.0   0.0
prior         0.25    0.25    0.25    0.25      
alpha         1.50


Literal listener


1ученик  2ученик  3ученик  4ученик
география      1.0      0.0      0.0      0.0
химия          0.0      0.5      0.5      0.0
алгебра        0.0      0.0      1.0      0.0
физика         0.0      0.0      0.0      1.0


Pragmatic speaker


география     химия   алгебра  физика
1ученик        1.0  0.000000  0.000000     0.0
2ученик        0.0  1.000000  0.000000     0.0
3ученик        0.0  0.261204  0.738796     0.0
4ученик        0.0  0.000000  0.000000     1.0


Pragmatic listener


1ученик   2ученик   3ученик  4ученик
география      1.0  0.000000  0.000000      0.0
химия          0.0  0.792893  0.207107      0.0
алгебра        0.0  0.000000  1.000000      0.0
физика         0.0  0.000000  0.000000      1.0

In [30]:
alpha_mod = RSA(lexicon=lex, prior=[0.25, 0.25, 0.25, 0.25], costs=[0.0, 0.0, 0.0, 0.0], alpha=4)
print("="*70 + "\nEven priors and all-0 message costs; alpha = 2\n")

display_reference_game(alpha_mod)

print("\nLiteral listener")
display(alpha_mod.literal_listener())

print("\nPragmatic speaker")
display(alpha_mod.speaker())

print("\nPragmatic listener")
display(alpha_mod.listener())

Even priors and all-0 message costs; alpha = 2



1ученик 2ученик 3ученик 4ученик costs
география     1.00     0.0     0.0     0.0   0.0
химия         0.00     1.0     1.0     0.0   0.0
алгебра       0.00     0.0     1.0     0.0   0.0
физика        0.00     0.0     0.0     1.0   0.0
prior         0.25    0.25    0.25    0.25      
alpha         4.00


Literal listener


1ученик  2ученик  3ученик  4ученик
география      1.0      0.0      0.0      0.0
химия          0.0      0.5      0.5      0.0
алгебра        0.0      0.0      1.0      0.0
физика         0.0      0.0      0.0      1.0


Pragmatic speaker


география     химия   алгебра  физика
1ученик        1.0  0.000000  0.000000     0.0
2ученик        0.0  1.000000  0.000000     0.0
3ученик        0.0  0.058824  0.941176     0.0
4ученик        0.0  0.000000  0.000000     1.0


Pragmatic listener


1ученик   2ученик   3ученик  4ученик
география      1.0  0.000000  0.000000      0.0
химия          0.0  0.944444  0.055556      0.0
алгебра        0.0  0.000000  1.000000      0.0
физика         0.0  0.000000  0.000000      1.0

А что, если изменить еще сильнее?

In [31]:
alpha_mod = RSA(lexicon=lex, prior=[0.25, 0.25, 0.25, 0.25], costs=[0.0, 0.0, 0.0, 0.0], alpha=9)
print("="*70 + "\nEven priors and all-0 message costs; alpha = 2\n")

display_reference_game(alpha_mod)

print("\nLiteral listener")
display(alpha_mod.literal_listener())

print("\nPragmatic speaker")
display(alpha_mod.speaker())

print("\nPragmatic listener")
display(alpha_mod.listener())

Even priors and all-0 message costs; alpha = 2



1ученик 2ученик 3ученик 4ученик costs
география     1.00     0.0     0.0     0.0   0.0
химия         0.00     1.0     1.0     0.0   0.0
алгебра       0.00     0.0     1.0     0.0   0.0
физика        0.00     0.0     0.0     1.0   0.0
prior         0.25    0.25    0.25    0.25      
alpha         9.00


Literal listener


1ученик  2ученик  3ученик  4ученик
география      1.0      0.0      0.0      0.0
химия          0.0      0.5      0.5      0.0
алгебра        0.0      0.0      1.0      0.0
физика         0.0      0.0      0.0      1.0


Pragmatic speaker


география     химия   алгебра  физика
1ученик        1.0  0.000000  0.000000     0.0
2ученик        0.0  1.000000  0.000000     0.0
3ученик        0.0  0.001949  0.998051     0.0
4ученик        0.0  0.000000  0.000000     1.0


Pragmatic listener


1ученик   2ученик   3ученик  4ученик
география      1.0  0.000000  0.000000      0.0
химия          0.0  0.998054  0.001946      0.0
алгебра        0.0  0.000000  1.000000      0.0
физика         0.0  0.000000  0.000000      1.0

Альфа - это параметр, который создает рациональность говорящего. Если сделать говорящего сильно рациональным, то настанет момент, когда первое значение максимально достигнет единицы, а другой - нуля (0.998054 и 0.001946). Какой вывод можем сделать? Сильно увеличивать альфа не стоит, так как чем выше альфа, тем "однозначнее" выбор.

Let's change the cost! (с изменением исходных (априорных) вероятностей объектов)

In [39]:
print("="*70 + "\nEven priors, imbalanced message costs\n")
cost_most = RSA(lexicon=lex, prior=[0.3, 0.2, 0.1, 0.4], costs=[-1.0, -1.0, 0.0, 3.0])

display_reference_game(cost_most)

print("\nLiteral listener")
display(cost_most.literal_listener())

print("\nPragmatic speaker")
display(cost_most.speaker())

print("\nPragmatic listener")
display(cost_most.listener())

Even priors, imbalanced message costs



1ученик 2ученик 3ученик 4ученик costs
география      1.0     0.0     0.0     0.0  -1.0
химия          0.0     1.0     1.0     0.0  -1.0
алгебра        0.0     0.0     1.0     0.0   0.0
физика         0.0     0.0     0.0     1.0   3.0
prior          0.3     0.2     0.1     0.4      
alpha          1.0


Literal listener


1ученик   2ученик   3ученик  4ученик
география      1.0  0.000000  0.000000      0.0
химия          0.0  0.666667  0.333333      0.0
алгебра        0.0  0.000000  1.000000      0.0
физика         0.0  0.000000  0.000000      1.0


Pragmatic speaker


география     химия   алгебра  физика
1ученик        1.0  0.000000  0.000000     0.0
2ученик        0.0  1.000000  0.000000     0.0
3ученик        0.0  0.109232  0.890768     0.0
4ученик        0.0  0.000000  0.000000     1.0


Pragmatic listener


1ученик   2ученик   3ученик  4ученик
география      1.0  0.000000  0.000000      0.0
химия          0.0  0.948213  0.051787      0.0
алгебра        0.0  0.000000  1.000000      0.0
физика         0.0  0.000000  0.000000      1.0

Тут мы опять наблюдаем, как исчезает импликатура, а выбор становится однозначнее. 

Интересный факт: минусовые значения могут быть только у цены, а вот alpha и prior минусовыми быть не могут (я так понимаю, это из-за того, что в формуле стоит экспонента, которая не может быть отрицательной).